### PR TITLE
Fix path of the states in `SeedGenerator` and tracking of `torch_params`

### PR DESCRIPTION
**Pull request by james77777778**
This PR:
- Fix the incorrect paths of the states in `SeedGenerator`
- Add `seed_generators` attr to `Layer` (primarily for testing purposes)
- Add assertion for len of `torch_params` in torch backend
- Update `torch_params` in `_track_variable` and `_untrack_variable`

In the current codebase, the fact that all states in the seed generator share an identical path `seed_generator_state` should be considered an issue.

Additionally, I have found some incorrect `expected_num_seed_generators` in `run_layer_test` within certain tests.

### CODE REVIEW

File: keras/layers/regularization/dropout.py [hunk 1]
```diff
@@ -49,10 +49,8 @@ def __init__(self, rate, noise_shape=None, seed=None, **kwargs):
         self.rate = rate
         self.seed = seed
         self.noise_shape = noise_shape
-        if rate > 0:
-            self.seed_generator = backend.random.SeedGenerator(seed)
+        self.seed_generator = backend.random.SeedGenerator(seed)
```
Comment (james77777778) [inline keras/layers/regularization/dropout.py:52]: I think we should always initialize a `self.seed_generator` regardless of how `rate` is set.
Otherwise, the user won't be able to reactivate the dropout layer by setting a new `rate` and recompiling the training graph.

Comment (fchollet) [reply keras/layers/regularization/dropout.py:52]: The rate is not intended to be settable. If we want to make it settable we should make a setter that would create the state variable (after unlocking the layer).

File: keras/random/seed_generator.py [hunk 2]
```diff
@@ -44,7 +46,12 @@ def call(self, x, training=False):
     ```
     """
 
-    def __init__(self, seed=None, **kwargs):
+    def __init__(self, seed=None, name=None, **kwargs):
+        if name is None:
+            name = auto_name(self.__class__.__name__)
+        self.name = name
+        self._parent_path = None
```
Comment (fchollet) [inline keras/random/seed_generator.py:53]: This isn't needed

File: keras/random/seed_generator.py [hunk 3]
```diff
@@ -66,13 +73,14 @@ def seed_initializer(*args, **kwargs):
...
             return self.backend.convert_to_tensor([seed, 0], dtype=dtype)
 
-        self.state = self.backend.Variable(
-            seed_initializer,
-            shape=(2,),
-            dtype="uint32",
-            trainable=False,
-            name="seed_generator_state",
-        )
+        with self._open_name_scope():
```
Comment (fchollet) [inline keras/random/seed_generator.py:76]: This isn't needed, instead you can just do `with backend.name_scope(self.name, caller=self):`

File: keras/layers/layer.py [hunk 4]
```diff
@@ -646,6 +648,22 @@ def metrics_variables(self):
                 vars.extend(metric.variables)
         return vars
 
+    @property
+    def seed_generators(self):
```
Comment (fchollet) [inline keras/layers/layer.py:652]: It isn't intended to be accessed by users, let's not make it public. If a user needs to access the RNG state variable for a specific layer they could it via public APIs, like `layer.seed_generator.state`.

File: keras/backend/torch/layer.py [hunk 5]
```diff
@@ -13,8 +13,8 @@ def _post_build(self):
         self._track_variables()
 
     def _track_variables(self):
-        self.torch_params = torch.nn.ParameterList(
-            [variable.value for variable in self.variables]
+        self.torch_params = torch.nn.ParameterDict(
```
Comment (fchollet) [inline keras/backend/torch/layer.py:16]: The dict is more readable, but it's unsafe -- there's no guarantee that variable names are unique within a model (except for Functional models). It's entirely possible to create models with duplicate variable paths, which would cause tracking issues above. So the list is preferable.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

Uniquifying paths for seed generator states is a good idea.

File: keras/layers/regularization/dropout.py (hunk 1 above)
Comment (james77777778) [reply keras/layers/regularization/dropout.py]: Got it. I have also applied this idea to `AlphaDropout`, `GaussianDropout` and `GaussianNoise`

File: keras/random/seed_generator.py (hunk 3 above)
Comment (james77777778) [reply keras/random/seed_generator.py]: Fixed

File: keras/layers/layer.py (hunk 4 above)
Comment (james77777778) [reply keras/layers/layer.py]: Got it. I have removed it and used an internal method `get_seed_generators`  in `testing.test_case.py` to get the correct number of seed generators.

File: keras/random/seed_generator.py (hunk 2 above)
Comment (james77777778) [reply keras/random/seed_generator.py]: Fixed

File: keras/backend/torch/layer.py (hunk 5 above)
Comment (james77777778) [reply keras/backend/torch/layer.py:17]: It should be safe to use `id(variable)` as a key for `torch.nn.ParameterDict`. I believe `BaseOptimizer` adopts the same approach to get the mapping of variables.

I could not find a solution for safely adding/removing the variable from `torch.nn.ParameterList`
(`in` and `remove` of `KerasVariable` are not supported)

Comment (fchollet) [reply keras/backend/torch/layer.py:17]: What's the original issue with using a list though?

Comment (james77777778) [reply keras/backend/torch/layer.py:17]:
```python
from keras import layers

layer = layers.Dense(units=16)
layer.build([None, 8])
assert len(layer.torch_params) == len(layer.variables)

layer.enable_lora(rank=2)
assert len(layer.torch_params) == len(layer.variables)  # <--

```
|`master`|this pr|
|-|-|
|Failed at line 8|Success|

The biggest issue is that `zero_grad()` will fail to reset the uncaptured variables.
I believe the currect LoRA implementation will not be trained correctly on torch backend.

Comment (james77777778) [reply keras/backend/torch/layer.py:17]: To demonstrate the error of `zero_grad` using LoRA:
```python
import torch

from keras import layers

layer = layers.Dense(units=3)
layer.build([None, 2])
layer.enable_lora(rank=1)


def list_grads(layer):
    grads = dict()
    for v in layer.trainable_variables:
        grads[v.name] = v.value.grad
    return grads


# Fake grads
for v in layer.trainable_variables:
    v.value.grad = torch.rand_like(v.value)

print(list_grads(layer))
layer.zero_grad()
print(list_grads(layer))

```

```bash
# master branch
{'bias': tensor([0.6259, 0.4827, 0.6012], device='cuda:0'), 'lora_kernel_a': tensor([[0.6620],
        [0.7231]], device='cuda:0'), 'lora_kernel_b': tensor([[0.7123, 0.9257, 0.1676]], device='cuda:0')}
{'bias': None, 'lora_kernel_a': tensor([[0.6620],
        [0.7231]], device='cuda:0'), 'lora_kernel_b': tensor([[0.7123, 0.9257, 0.1676]], device='cuda:0')}

# this pr
{'bias': tensor([0.5960, 0.2336, 0.1569], device='cuda:0'), 'lora_kernel_a': tensor([[0.9123],
        [0.9217]], device='cuda:0'), 'lora_kernel_b': tensor([[0.3435, 0.9276, 0.6599]], device='cuda:0')}
{'bias': None, 'lora_kernel_a': None, 'lora_kernel_b': None}
```

File: keras/backend/torch/layer.py [hunk 6]
```diff
@@ -13,8 +13,9 @@ def _post_build(self):
         self._track_variables()
 
     def _track_variables(self):
-        self.torch_params = torch.nn.ParameterList(
-            [variable.value for variable in self.variables]
+        # Index given to ParameterDict must be a string
+        self.torch_params = torch.nn.ParameterDict(
+            {str(id(variable)): variable.value for variable in self.variables}
```
Comment (fchollet) [inline keras/backend/torch/layer.py:18]: There should be no need to cast to `str` since `id` is hashable?

File: keras/layers/layer.py [hunk 7]
```diff
@@ -1204,13 +1204,24 @@ def _track_variable(self, variable):
         if not self.trainable:
             variable.trainable = False
 
+        if backend.backend() == "torch" and hasattr(self, "torch_params"):
```
Comment (fchollet) [inline keras/layers/layer.py:1207]: Please move these 2 blocks to the torch Layer class. Just override the method, call super() and then add this logic. For both `_track_variable` and `_untrack_variable`. As a general rule we should strictly avoid backend-specific code in the main codebase.

File: keras/layers/preprocessing/discretization.py [hunk 8]
```diff
@@ -147,7 +147,6 @@ def __init__(
         self.sparse = sparse
 
         if self.bin_boundaries:
-            self.built = True
```
Comment (fchollet) [inline keras/layers/preprocessing/discretization.py:150]: Why this change?

File: keras/layers/regularization/dropout.py [hunk 9]
```diff
@@ -52,7 +52,6 @@ def __init__(self, rate, noise_shape=None, seed=None, **kwargs):
         if rate > 0:
             self.seed_generator = backend.random.SeedGenerator(seed)
         self.supports_masking = True
-        self.built = True
```
Comment (fchollet) [inline keras/layers/regularization/dropout.py:55]: Why this change?

**Review by fchollet — COMMENTED** [review]
Thanks for the update!

File: keras/layers/layer.py (hunk 7 above)
Comment (james77777778) [reply keras/layers/layer.py]: Done

File: keras/layers/preprocessing/discretization.py (hunk 8 above)
Comment (james77777778) [reply keras/layers/preprocessing/discretization.py:150]: Explicitly set `self.built=True` in `__init__` will fail the `run_build_asserts` of symbolic call test in `run_layer_test`.
The root cause is that `self.torch_params` will not be initialized.

I'm unsure where the issue is but it should be acceptable to leave `self.built=True` for `build` method.

File: keras/layers/regularization/dropout.py (hunk 9 above)
Comment (james77777778) [reply keras/layers/regularization/dropout.py:55]: Same as above

File: keras/layers/preprocessing/discretization.py (hunk 8 above)
Comment (james77777778) [reply keras/layers/preprocessing/discretization.py:150]:
```python
from keras import layers

inputs = layers.Input([2])
layer = layers.Dropout(rate=0.2)
layer(inputs)
assert len(layer.torch_params) == len(layer.variables)

```

This script will fail at the master branch but work fine in this pr.

File: keras/ops/operation.py [hunk 10]
```diff
@@ -280,3 +280,11 @@ def _post_build(self):
     def _setattr_hook(self, name, value):
         """Can be overridden for per backend post build actions."""
         return name, value
+
+    def _post_track_variable(self, variable):
+        """Can be overridden for per backend post track actions."""
+        pass
+
+    def _post_untrack_variable(self, variable):
```
Comment (james77777778) [inline keras/ops/operation.py:288]: Two hooks have been introduced to enable the postprocessing of `_track_variable` and `_untrack_variable` in torch backend

File: keras/layers/layer.py (hunk 7 above)
Comment (james77777778) [reply keras/layers/layer.py]: Fixed now.

**Review by fchollet — APPROVED** [review]
LGTM, thank you!
